File: common/scala/src/main/scala/org/aiddl/common/scala/math/linear_algebra/LupSolver.scala
package org.aiddl.common.scala.math.linear_algebra

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.representation._
import scala.collection.mutable.HashMap

import org.aiddl.common.scala.math.linear_algebra.Matrix
import org.aiddl.common.scala.math.linear_algebra.AiddlMatrix
import org.aiddl.common.scala.math.linear_algebra.LupDecomposition

import org.aiddl.core.scala.representation.TermImplicits.term2Num
import org.aiddl.core.scala.representation.TermImplicits.double2Num


class LupSolver extends Function {
    def apply( x: Term ): Term = ???

    def apply( l: Matrix, u: Matrix, pi: Matrix, b: Matrix ): Matrix = {
        val x = Array.fill(l.m)(Num(0))
        val y = Array.fill(l.m)(Num(0))

        (0 until l.m).foreach( i => {
            val s = (0 until i).foldLeft(Num(0))( (c, j) => c + l(i, j) * y(j) )
            val y_i = b(pi(i).asInt.x.intValue) - s
            y.update(i, y_i)
        })

        (l.m-1 to 0 by -1).foreach( i => {
            val s = (i+1 until l.m).foldLeft(Num(0))( (c, j) => c + u(i, j) * x(j) )
            val x_i = (y(i) - s) / u(i, i)
            x.update(i, x_i)
        })
        AiddlMatrix.vec(x.toSeq: _*)
    }
}   

File: common/java/src/main/java/org/aiddl/common/java/math/linear_algebra/LupSolver.java
package org.aiddl.common.java.math.linear_algebra;

import java.util.ArrayList;
import java.util.List;

import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.Term;

public class LupSolver implements Function {

	@Override
	public Term apply(Term args) {
		Term L = args.get(0);
		Term U = args.get(1);
		Term pi = args.get(2);
		Term b = args.get(3);
			
		Integer n = L.size();
		
		List<Term> x = new ArrayList<>();
		List<Term> y = new ArrayList<>();
		for ( int i = 0 ; i < n ; i++ ) {
			x.add(Term.real(0.0));
			y.add(Term.real(0.0));
		}
		
		for ( int i = 0 ; i < n ; i++ ) {
			double sum = 0.0;
			for ( int j = 0 ; j < i ; j++ ) {
				sum += L.get(i).get(j).getDoubleValue() * y.get(j).getDoubleValue();
			}
			double y_i = b.get(pi.get(i).getIntValue()).getDoubleValue() - sum;
			y.set(i, Term.real(y_i));
		}
		
		for ( int i = n-1; i >= 0 ; i-- ) {
			double sum = 0.0;
			for ( int j = i+1 ; j < n ; j++ ) {
				sum += U.get(i).get(j).getDoubleValue() * x.get(j).getDoubleValue();
			}
			double x_i = (y.get(i).getDoubleValue() - sum) / U.get(i).get(i).getDoubleValue();
			x.set(i, Term.real(x_i));
		}
		
		return Term.tuple(x);

	}
}
